Flaky test: TestRunAttachDetachFromInvalidFlag

### Description

Doesn't appear to be failing often, but has failed a couple of times over the
years, and may be racy; most recently failed on https://github.com/moby/moby/pull/44126

```
=== FAIL: amd64.integration-cli TestDockerCLIRunSuite/TestRunAttachDetachFromInvalidFlag (0.85s)
    docker_cli_run_unix_test.go:233: read |0: file already closed
    --- FAIL: TestDockerCLIRunSuite/TestRunAttachDetachFromInvalidFlag (0.85s)
```


- https://github.com/moby/moby/pull/36424#issuecomment-368806539
- https://github.com/moby/moby/pull/34554#issuecomment-326156589
- https://github.com/moby/moby/pull/34707#issuecomment-327257476
- https://github.com/moby/moby/pull/31211#issuecomment-281320818
- https://github.com/moby/moby/pull/39683#issuecomment-519180161

--- a/integration-cli/docker_cli_run_unix_test.go
+++ b/integration-cli/docker_cli_run_unix_test.go
@@ -205,41 +205,6 @@ func (s *DockerCLIRunSuite) TestRunAttachDetachFromFlag(c *testing.T) {
 	assert.Equal(c, running, "true", "expected container to still be running")
 }
 
-// TestRunAttachDetachFromInvalidFlag checks attaching and detaching with the escape sequence specified via flags.
-func (s *DockerCLIRunSuite) TestRunAttachDetachFromInvalidFlag(c *testing.T) {
-	const name = "attach-detach"
-	cli.DockerCmd(c, "run", "--name", name, "-itd", "busybox", "top")
-	cli.WaitRun(c, name)
-
-	// specify an invalid detach key, container will ignore it and use default
-	cmd := exec.Command(dockerBinary, "attach", "--detach-keys=ctrl-A,a", name)
-	stderr, err := cmd.StderrPipe()
-	if err != nil {
-		c.Fatal(err)
-	}
-	cpty, tty, err := pty.Open()
-	if err != nil {
-		c.Fatal(err)
-	}
-	defer cpty.Close()
-	cmd.Stdin = tty
-	if err := cmd.Start(); err != nil {
-		c.Fatal(err)
-	}
-	go cmd.Wait()
-
-	bufReader := bufio.NewReader(stderr)
-	out, err := bufReader.ReadString('\n')
-	if err != nil {
-		c.Fatal(err)
-	}
-	// it should print a warning to indicate the detach key flag is invalid
-	// FIXME(thaJeztah): this is a regression: current versions of docker (cli) don't print the error message
-	// errStr := "Invalid detach keys (ctrl-A,a) provided"
-	errStr := "unable to upgrade to tcp, received 400"
-	assert.Equal(c, strings.TrimSpace(out), errStr)
-}
-
 // TestRunAttachDetachFromConfig checks attaching and detaching with the escape sequence specified via config file.
 func (s *DockerCLIRunSuite) TestRunAttachDetachFromConfig(c *testing.T) {
 	keyCtrlA := []byte{1}

--- a/integration/container/attach_test.go
+++ b/integration/container/attach_test.go
@@ -1,6 +1,8 @@
 package container
 
 import (
+	"net/http"
+	"net/url"
 	"testing"
 	"time"
 
@@ -11,6 +13,7 @@ import (
 	systemutil "github.com/moby/moby/v2/integration/internal/system"
 	"github.com/moby/moby/v2/internal/testutil"
 	"github.com/moby/moby/v2/internal/testutil/daemon"
+	"github.com/moby/moby/v2/internal/testutil/request"
 	"gotest.tools/v3/assert"
 	is "gotest.tools/v3/assert/cmp"
 	"gotest.tools/v3/poll"
@@ -61,6 +64,40 @@ func TestAttach(t *testing.T) {
 			assert.Check(t, is.Equal(mediaType, tc.expectedMediaType))
 		})
 	}
+}
+
+func TestAttachRejectsInvalidDetachKeys(t *testing.T) {
+	ctx := setupTest(t)
+	apiClient := testEnv.APIClient()
+
+	resp, err := apiClient.ContainerCreate(ctx, client.ContainerCreateOptions{
+		Config: &container.Config{
+			Image: "busybox",
+			Cmd:   []string{"top"},
+			Tty:   true,
+		},
+		HostConfig:       &container.HostConfig{},
+		NetworkingConfig: &network.NetworkingConfig{},
+	})
+	assert.NilError(t, err)
+
+	query := url.Values{"detachKeys": {"ctrl-A,a"}}
+	res, body, err := request.Post(ctx, "/containers/"+resp.ID+"/attach?"+query.Encode(),
+		request.With(func(req *http.Request) error {
+			req.Header.Set("Connection", "Upgrade")
+			req.Header.Set("Upgrade", "tcp")
+			return nil
+		}),
+	)
+	if body != nil {
+		defer func() { _ = body.Close() }()
+	}
+	assert.NilError(t, err)
+	assert.Equal(t, res.StatusCode, http.StatusBadRequest)
+
+	out, err := request.ReadBody(body)
+	assert.NilError(t, err)
+	assert.Check(t, is.Contains(string(out), "Invalid detach keys"))
 }
 
 // Regression test for #37182

